drivers: serial: nrfx_uarte: Add support for higher baudrates

nordic-krch opened the pull request:
Fast instances (uart00 in nRF54Lx and uart120 in nRF54H20) support >1M baudates. Extend driver to support that.

Updated overlay for nrf54h20dk in `uart_async_api` test to use 4Mbaud for uart120 instance. This shown an issue in the test which was not ready for higher baudrates. Test was tweaked to adjust its timings based on used baudrate.

nordic-krch committed:
dts: bindings: serial: nordic,nrf-uart-common: Extend baudrate enum

Add faster baudrates (2M, 4M and 8M).

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

nika-nordic commented on dts/bindings/serial/nordic,nrf-uart-common.yaml:
```diff
@@ -43,3 +43,6 @@ properties:
       - 460800
       - 921600
       - 1000000
+      - 2000000
```
shouldn't we introduce `nordic,nrf-uart-fast.yaml` or similar to differentiate between devices that have >1Mbaud capabilities? Take a look at this PR for reference: https://github.com/zephyrproject-rtos/zephyr/pull/79330

nordic-krch replied:
Driver compilation will fail if not supported `current-speed` is set for an instance. Alternative approach would be to use separate compatible but this will trigger more work as currently `current-speed` is an enum in `nordic,nrf-uart-common.yaml` and that would need to be extracted because enums in bindings cannot be overwritten. Additionally, support for two compatibles would need be added in places like: https://github.com/zephyrproject-rtos/zephyr/blob/c3f979b7367f892ec4ec63b9ebb102810e910050/drivers/pinctrl/pinctrl_nrf.c#L45. I'm not sure what we gain apart from error detected in DT vs error detected during compilation.

bjarki-andreasen replied:
If we where completely true to the devicetree, if its the "same" underlying hardware block, and "fast" just means, "supports higher bitrate", we would add a flag/boolean to the common binding like:
```
uart@xxx {
        compatible = "nordic,nrf-uarte"; /* same hardware, targets same driver */
        fast; /* <this instance is fast */
};
```
the enum should contain a superset of all supported speeds, for all variations, including fast, the driver then validates `current-speed` based on if the `fast;` flag exists. DT really is not there to validate stuff like `current-speed`, `current-speed` is a zephyr-ism we use to set initial configs, DT enums are just a "base filter"

bjarki-andreasen replied:
Consider how many `nordic,nrf-comp` variations there would be if we had one compat if it supported low power mode, one if it supported some specific subset of voltage references etc. pure chaos :)

nordic-krch replied:
`fast` property is currently redundant because fast instance can be determined based on clock frequency that drives it (16MHz - slow, >16MHz fast, currently 128M in nrf54l and 320M in nrf54h). Based on frequency that clocks the uart at compile time there is a build time check if `current-speed` can be achieved.

bjarki-andreasen replied:
Perfect, achieves the same thing in this case if the clock can not be selected at runtime :) Though a runtime check would also be needed if UART can be configured runtime :)

nordic-krch replied:
runtime check is also there.

nordic-krch committed:
drivers: serial: nrfx_uarte: Rename FAST to FAST_PD

Rename UARTE_ANY_FAST to UARTE_ANY_FAST_PD. There are 2 types of
'fast' UARTE instances. In nrf54h20 instance uart120 is in fast
power domain that requires additional power and clock management
of that domain. In nrf54lx fast uart00 instance does not require
that. Add _PD to indicate fast power domain.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
drivers: serial: nrfx_uarte: Add support for higher baudrates

nrf54x devices have UARTE instance capable of using baudrate higher
than 1M. Higher baudrates does not have predefined values for
BAUDRATE register. A formula can be used to calculate BAUDRATE
value that shall be used for desired baudrate. Add UARTE_ANY_HIGH_SPEED
macro which is set when high speed is enabled (uart00 in nrf54lx or
uart120 in nrf54h20). For high speed instance use formula for getting
value that shall be written to BAUDRATE register.

When runtime configuration is not used then same formula is used to
calculate fixed BAUDRATE value.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
drivers: serial: nrfx_uarte: Fix spurious RXTO event

Fast instance in nrf54h (uart120) can generate a spurious RXTO event
some time after RXTO event that indicates that RX path is disabled.
The time when event is generated depends on baudrate and when slower
baudrates are used peripheral is disabled on time to not notice it
in the test but with higher baudates issue become visible. In order
to avoid spurious interrupt, RXTO interrupt is disabled during RXTO
event handling and enabled when RX is enabled. This workaround is
applied only for fast instance to avoid unnecessary register
accesses for slower instances.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
tests: drivers: uart: uart_async_api: Tweak test_read_abort test

Test was staring a TX transfer and aborting it after 300 us from
the timer handler. Test was assuming that ongoing transfer will
not finish in those 300 us. This might not be true for higher
baudrates. Instead of using fixed timeout, a value is calculated
from the baudrate and targets to abort the transfer after approx.
20 bytes of 95 byte long transfer.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
tests: drivers: uart: uart_async_api: Use 4Mbaud for nrf54h20dk uart120

Use high baudrate when testing uart120 on nrf54h20dk/nrf54h20/cpuapp.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

nordic-krch commented:
@anangl can you take a look?
